storage/repair.go

```diff
@@ -124,23 +128,280 @@ func (r shardRepairer) Repair(
 	// Add peer metadata
 	peerIter, err := session.FetchBlocksMetadataFromPeers(namespace, shard.ID(), start, end)
 	if err != nil {
-		return repair.MetadataComparisonResult{}, err
+		return repair.Result{}, err
 	}
 	if err := metadata.AddPeerMetadata(peerIter); err != nil {
-		return repair.MetadataComparisonResult{}, err
+		return repair.Result{}, err
 	}
 
 	metadataRes := metadata.Compare()
+	result := repair.Result{DifferenceSummary: metadataRes}
 
-	r.recordFn(namespace, shard, metadataRes)
+	res, err := r.repairFn(namespace, shard, metadataRes)
+	if err != nil {
+		// even in the event of repair failure, report statistics about differences
+		r.recordFn(namespace, shard, result)
+		return result, err
+	}
 
-	return metadataRes, nil
+	result.RepairSummary = res
+	r.recordFn(namespace, shard, result)
+	return result, nil
 }
 
-func (r shardRepairer) recordDifferences(
+type hostSet map[string]repair.HostBlockMetadata
+
+type blockID struct {
+	idHash ts.Hash
+	start  time.Time
+}
+
+func newHostSet() *hostSet {
+	hs := make(hostSet, 3)
+	return &hs
+}
+
+func (h *hostSet) insert(hbm repair.HostBlockMetadata) {
+	if host := hbm.Host; host != nil {
+		(*h)[host.String()] = hbm
+	}
+}
+
+func (h *hostSet) insertOneFromSet(oh *hostSet) {
+	for _, hbm := range *oh {
+		h.insert(hbm)
+		return
+	}
+}
+
+func (h *hostSet) contains(host topology.Host) bool {
+	if host == nil {
+		return false
+	}
+	_, ok := (*h)[host.String()]
+	return ok
+}
+
+func (h *hostSet) remove(host topology.Host) {
+	delete(*h, host.String())
```

Comment:
same, value type should be enough. Also do we need the nil check as in `contains`?